### ISSUE TITLE
Identify user mistakes in config

### ISSUE THREAD

**Issue by antonmedv** [deployphp/deployer#976]
| Q                 | A
| ----------------- | ---
| Issue Type        | Feature Request
| Deployer Version  | next

Implement internal bookkeeping to help identify user mistakes like spelling errors or unused variables that may lead to unexpected behavior.

```php
set('keep_relaeses', 5);
```

Shows an warning:

```
Unknown config `keep_relaeses`. Did you mean `keep_releases`?
```

**Comment by pluseg** [comment]
@elfet As I understand docs (https://deployer.org/docs/configuration), user can define his custom params. So he would hypothetically set `keep_relaeses` param. Could you please provide more details about this feature? Would you like to have list of reserved configs or add spelling check in get() function?

**Comment by elfet** [comment]
I think it should be predefined list of configuration option.

**Comment by pluseg** [comment]
@elfet excuse me, but I still don't understand the mechanics of the feature. Assume we have predefined list of `$predefined = ['shared_files', 'shared_dirs', 'keep_releases', ...]` and try to set any "wrong" param, f.e. `keep_release`. Do we need to check into `set()` function? And how can we define if `keep_release` is wrong or not? Maybe user wants to define both `keep_releases` and `keep_release` variables.

**Comment by samplenull** [comment]
Another way is to add constants for predefined variables like Config::VAR_SHARED_FILES, but it is not good for readability of configuration.

**Comment by brunofontes** [comment]
Hello! This is my first time contributing to something, so sorry if I am saying a very dumb thing, but how about if it check the set values and, if it is not a predefined one, it checks if the value is being called in the config file?

**Comment by antonmedv** [comment]
Currently, the idea to add validation callback:
```php
set('keep_releases', 5, function ($value) {
    return is_number($value);
});
```

**Comment by brunofontes** [comment]
I was also thinking about a quick compare and, if it were close enough of the predefined ones, it would emit a warning. Maybe adding a flag to avoid the warning if the user really wants to use a 'keep_relases' or similar, but that would not validate the data.

Using the validation callback looks interesting. But if the user write it, we would need to check the `$name` anyway and provide some default callback to the predefined configurations, wouldn't we?

Thanks!

**Comment by antonmedv** [comment]
Yes, all config in common.php will need some callbacks.

**Comment by brunofontes** [comment]
I can help on this callback function. I am just thinking about a way to check them both, `$name` and `$value` without needing to include a flag in the `set` function.

**Comment by antonmedv** [comment]
I think, check for the name can be something like this:
* If user made mistake in one letter of any config: `keeb_releases`
* Check using (for example via levengshtaine distance) nearest look alike config
* Show warning, just for information purposes. Maybe everything is okay 👌

**Comment by brunofontes** [comment]
That might work. I will give it a try. :)

**Comment by brunofontes** [comment]
I had some issues on my side here but I am finishing the first part (validate the `$value` with the callback function). Do you want to take a look at that before I include the `$name` validation? If so, how do you prefer to do it?

**Comment by antonmedv** [comment]
Sure, make pr please

**Comment by brunofontes** [comment]
As I said, this is my first try in a collaborative work, so just let me know if it is really bad and I can try to fix it.

Thanks!

**Comment by brunofontes** [comment]
I also included the `$name` verification with the levenshtaine distance, but I am not very sure about the formating of the warning message.

Thanks!

**Comment by antonmedv** [comment]
```
❯ dep deploy -o repositor=1
warning "pwd" not found in parameters. Did you mean "env"?
warning "shared_fles" not found in parameters. Did you mean "shared_files"?
warning "pwd" not found in parameters. Did you mean "env"?
warning "shared_fles" not found in parameters. Did you mean "shared_files"?
^C⠇ connect deployer.pw
```
Looks not really good =/
I'll try to improve it.

**Comment by antonmedv** [comment]
Fixed by 76f07b4e05b898db852686ef8cac93d8b1a7f692
<img width="505" alt="Снимок экрана 2020-11-04 в 12 39 22" src="https://user-images.githubusercontent.com/141232/98153282-ff072580-1ee3-11eb-983f-3f8800bad78b.png">
